### PR TITLE
Closable Pool

### PR DESCRIPTION
**Pull request by mosquito**

### CODE REVIEW

File: aio_pika/pool.py [hunk 1]
```diff
@@ -47,24 +48,45 @@ def _is_overflow(self) -> bool:
...
 
         return await self._create_item()
 
     def put(self, item: T):
+        if self.__closed:
+            raise asyncio.InvalidStateError('Pool closed')
+
         return self.__items.put_nowait(item)
 
+    async def close(self):
```
Comment (mosquito) [inline aio_pika/pool.py]: it's there now

**Comment by mosquito** [comment]
fixes #223

File: tests/test_pool.py [hunk 2]
```diff
@@ -60,6 +60,70 @@ class TestCase(BaseTestCase):
...
+                raise RuntimeError
+
+            self.closed = True
+
+    def setUp(self):
+        self.instances = set()
+        super().setUp()
+
+    async def create_instance(self):
+        obj = TestCaseClose.Instanse()
```
Comment (decaz) [inline tests/test_pool.py]: Instan**c**e

File: tests/test_pool.py [hunk 3]
```diff
@@ -60,6 +60,70 @@ class TestCase(BaseTestCase):
...
+        self.instances.add(obj)
+        return obj
+
+    async def test_close(self):
+        async def getter():
+            async with self.pool.acquire() as instance:
+                assert instance > 0
```
Comment (decaz) [inline tests/test_pool.py]: Constructor returns object of `Instance` class which doesn't have `__gt__` method.

File: tests/test_pool.py [hunk 4]
```diff
@@ -60,6 +60,70 @@ class TestCase(BaseTestCase):
...
+
+        await self.pool.close()
+
+        for instance in self.instances:
+            self.assertTrue(instance.closed)
+
+    async def test_close_context_manager(self):
+        async def getter():
+            async with self.pool.acquire() as instance:
+                assert instance > 0
```
Comment (decaz) [inline tests/test_pool.py]: The same as for `test_close`:
> Constructor returns object of `Instance` class which doesn't have `__gt__` method.

File: tests/test_pool.py [hunk 5]
```diff
@@ -60,6 +60,70 @@ class TestCase(BaseTestCase):
...
+        )
+
+        for instance in self.instances:
+            self.assertFalse(instance.closed)
+
+        await self.pool.close()
+
+        for instance in self.instances:
+            self.assertTrue(instance.closed)
+
```
Comment (decaz) [inline tests/test_pool.py]: `self.assertTrue(self.pool.is_closed)` can be added.

File: tests/test_pool.py [hunk 6]
```diff
@@ -60,6 +60,83 @@ class TestCase(BaseTestCase):
...
+            )
+
+            self.assertTrue(len(self.instances) > 1)
+
+            for instance in self.instances:
+                self.assertFalse(instance.closed)
+
+            self.assertFalse(self.pool.is_closed)
+
+        self.assertTrue(self.pool.is_closed)
```
Comment (decaz) [inline tests/test_pool.py]: Duplication I guess - there is equal assertion below.

File: tests/test_pool.py [hunk 7]
```diff
@@ -60,6 +60,83 @@ class TestCase(BaseTestCase):
         self.assertEqual(self.counter, self.max_size)
 
 
+class TestCaseClose(BaseTestCase):
```
[the remark this answers is not in the document]
Comment (mosquito) [reply tests/test_pool.py]: The ugly name just for tests collecting.

**Pull request merged**